src/physics/em/detail/UrbanMscScatter.hh

```diff
@@ -0,0 +1,532 @@
+        bool      small_step = (true_path < tsmall);
+        real_type theta0     = (small_step)
+                               ? std::sqrt(true_path / tsmall)
+                                     * this->compute_theta0(tsmall, end_energy)
+                               : this->compute_theta0(true_path, end_energy);
+
+        // Protect for very small angles
+        real_type theta2 = ipow<2>(theta0);
+        if (theta2 < params_.tau_small)
+        {
+            return cth;
+        }
+
+        if (theta0 > constants::pi / 6)
+        {
+            // theta0 > theta0_max
+            return this->simple_scattering(rng, xmean, x2mean);
+        }
+
+        real_type x = theta2 * (1 - theta2 / 12);
+        if (theta2 > real_type(0.01))
+        {
+            real_type sth = 2 * std::sin(real_type(0.5) * theta0);
+            x             = ipow<2>(sth);
+        }
+
+        // Evaluate parameters for the tail distribution
+        real_type lambdaeff = true_path / tau;
+        real_type ltau      = std::log(tau);
+        real_type u = small_step ? std::exp(std::log(tsmall / lambda) / 6)
+                                 : std::exp(ltau / 6);
+        real_type xx  = std::log(lambdaeff / rad_length_);
+        real_type xsi = msc_.coeffc1 + u * (msc_.coeffc2 + msc_.coeffc3 * u)
+                        + msc_.coeffc4 * xx;
+
+        // The tail should not be too big
+        xsi = max<real_type>(xsi, real_type(1.9));
+
+        real_type c = xsi;
+
+        if (std::abs(c - 3) < real_type(0.001))
+        {
+            c = real_type(3.001);
+        }
+        else if (std::abs(c - 2) < real_type(0.001))
+        {
+            c = real_type(2.001);
+        }
+
+        real_type c1 = c - 1;
+
+        real_type ea     = std::exp(-xsi);
+        real_type eaa    = 1 - ea;
+        real_type xmean1 = 1 - (1 - (1 + xsi) * ea) * x / eaa;
+        real_type x0     = 1 - xsi * x;
+
+        if (xmean1 <= real_type(0.999) * xmean)
+        {
+            cth = this->simple_scattering(rng, xmean, x2mean);
+        }
+
+        // From continuity of derivatives
+        real_type b1 = 2 + (c - xsi) * x;
+        real_type bx = c * x;
+
+        real_type eb1 = std::exp(std::log(b1) * c1);
+        real_type ebx = std::exp(std::log(bx) * c1);
+        real_type d   = ebx / eb1;
+
+        real_type xmean2 = (x0 + d - (bx - b1 * d) / (c - 2)) / (1 - d);
+
+        real_type f1x0 = ea / eaa;
+        real_type f2x0 = c1 / (c * (1 - d));
+        real_type prob = f2x0 / (f1x0 + f2x0);
+
+        real_type qprob = xmean / (prob * xmean1 + (1 - prob) * xmean2);
+
+        // Sampling of cos(theta)
+        if (generate_canonical(rng) < qprob)
+        {
+            real_type var = 0;
+            if (BernoulliDistribution(prob)(rng))
+            {
+                cth = 1 + std::log(ea + generate_canonical(rng) * eaa) * x;
+            }
+            else
+            {
+                var = (1 - d) * generate_canonical(rng);
```

Comment:
// Sample cos \theta from the distribution g_2(cos \theta)

Reply:
Added.